### PR TITLE
Fixed addr of bracket expression. Fixes #2148.

### LINKED ISSUE #2148: Streams module not working with JS backend
```nimrod
import streams

var x = newStringStream("foo")
```

```
lib/pure/streams.nim(222, 20) Error: internal error: expr(nkBracketExpr, tyChar)
Traceback (most recent call last)
nim.nim(92)              nim
nim.nim(56)              handleCmdLine
main.nim(269)            mainCommand
main.nim(112)            commandCompileToJS
modules.nim(203)         compileProject
modules.nim(151)         compileModule
passes.nim(193)          processModule
passes.nim(137)          processTopLevelStmt
jsgen.nim(1710)          myProcess
jsgen.nim(1700)          genModule
jsgen.nim(1561)          genStmt
jsgen.nim(1648)          gen
jsgen.nim(1222)          genVarStmt
jsgen.nim(1184)          genVarInit
jsgen.nim(1609)          gen
jsgen.nim(1053)          genCall
jsgen.nim(1571)          gen
jsgen.nim(1011)          genSym
jsgen.nim(1551)          genProc
jsgen.nim(1561)          genStmt
jsgen.nim(1642)          gen
jsgen.nim(1561)          genStmt
jsgen.nim(1657)          gen
jsgen.nim(832)           genAsgn
jsgen.nim(814)           genAsgnAux
jsgen.nim(1571)          gen
jsgen.nim(1011)          genSym
jsgen.nim(1551)          genProc
jsgen.nim(1561)          genStmt
jsgen.nim(1642)          gen
jsgen.nim(1561)          genStmt
jsgen.nim(1646)          gen
jsgen.nim(775)           genIf
jsgen.nim(1642)          gen
jsgen.nim(1561)          genStmt
jsgen.nim(1609)          gen
jsgen.nim(1054)          genCall
jsgen.nim(1048)          genArgs
jsgen.nim(1034)          genArg
jsgen.nim(1614)          gen
jsgen.nim(1450)          genConv
jsgen.nim(1615)          gen
jsgen.nim(972)           genAddr
msgs.nim(845)            internalError
msgs.nim(822)            liMessage
msgs.nim(724)            handleError
FAILURE
```

### PR DESCRIPTION
**Pull request by yglukhov**

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -976,14 +976,17 @@ proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
...
-    of tyArray, tyArrayConstr, tyOpenArray, tySequence, tyString, tyCString,
-       tyVarargs, tyChar:
-      genArrayAddr(p, n.sons[0], r)
-    of tyTuple:
-      genFieldAddr(p, n.sons[0], r)
-    else: internalError(n.sons[0].info, "expr(nkBracketExpr, " & $ty.kind & ')')
+    if ty.kind in MappedToObject:
+      gen(p, n.sons[0], r)
+    else:
+      let kindOfIndexedExpr = n.sons[0].sons[0].typ.kind
```
Comment (Araq) [inline compiler/jsgen.nim]: You really need this sing and dance with skipTypes(t, abstractVarRange) to access n.sons[0].sons[0] properly.

**Comment by yglukhov** [comment]
@Araq, I would be grateful for some example of a case which my change would break with for skipTypes reason. Not that I'm arguing. I just think I don't understand it as deeply.

**Comment by Araq** [comment]
Test it with a generic type that is instantiated like Foo[int]

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -976,14 +976,17 @@ proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
...
-    of tyArray, tyArrayConstr, tyOpenArray, tySequence, tyString, tyCString,
-       tyVarargs, tyChar:
-      genArrayAddr(p, n.sons[0], r)
-    of tyTuple:
-      genFieldAddr(p, n.sons[0], r)
-    else: internalError(n.sons[0].info, "expr(nkBracketExpr, " & $ty.kind & ')')
+    if ty.kind in MappedToObject:
+      gen(p, n.sons[0], r)
+    else:
+      let kindOfIndexedExpr = skipTypes(n.sons[0].sons[0].typ, abstractVarRange).kind
```
Comment (Araq) [inline compiler/jsgen.nim]: Still wrong. ;-) you need to call it twice. Or better yet, come up with a new helper 'proc skipTypes(x: PType; indexes: varargs[int]): PType'

**Comment by yglukhov** [comment]
Will do.

**Comment by yglukhov** [comment]
@Araq, I still don't get it. Here's a sample I test which works fine with this PR:
```nim
type Foo[T] = object
    a:T
type FooArr[T] = array[2, T]
var x: FooArr[Foo[int]]
var y = addr x[1]
y[].a = 5
echo x[1].a
```

**Comment by Araq** [comment]
Well it might be that something else already skips the tyGenericInst away then, but this doesn't make it right. It's fragile code when you don't check for it.

**Comment by yglukhov** [comment]
Thank you!

**Pull request merged**